plugins/algoltek-usb/fu-algoltek-usb-device.c

```diff
@@ -0,0 +1,567 @@
+
+		fu_byte_array_set_size(part_firmware, transfer_data_size, 0);
+
+		for (guint i = 0; i < transfer_data_size; i++)
+			part_firmware->data[i] = firmware_data->data[i + start_index];
+
+		if (!g_usb_device_control_transfer(usb_device,
+						   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+						   G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+						   G_USB_DEVICE_RECIPIENT_INTERFACE,
+						   ALGOLTEK_WRF,
+						   value, /* value */
+						   index, /* index */
+						   part_firmware->data,
+						   part_firmware->len,
+						   NULL,
+						   ALGOLTEK_DEVICE_USB_TIMEOUT,
+						   NULL,
+						   error)) {
+			g_prefix_error(error, "failed to contact device: ");
+			return FALSE;
+		}
+
+		start_index += transfer_data_size;
+		rest_transfer_data_size -= transfer_data_size;
+		if (count_check == (256 / max_packet_size) || rest_transfer_data_size == 0) {
+			count_check = 0;
+			if (!fu_algoltek_device_status_check(self, error))
+				return FALSE;
+		}
+		fu_progress_set_percentage_full(progress, start_index, firmware_data->len);
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_usb_device_setup(FuDevice *device, GError **error)
+{
+	FuAlgoltekUsbDevice *self = FU_ALGOLTEK_USB_DEVICE(device);
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(device));
+	g_autoptr(GByteArray) version_data = NULL;
+	g_autofree gchar *version_str = NULL;
+
+	/* UsbDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_algoltek_usb_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	version_data = algoltek_device_rdv(self, error);
+	version_str = g_strdup_printf("%s", version_data->data);
+
+	g_assert(self != NULL);
+	g_assert(usb_device != NULL);
+	fu_device_set_version(device, version_str);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_usb_device_write_firmware(FuDevice *device,
+				      FuFirmware *firmware,
+				      FuProgress *progress,
+				      FwupdInstallFlags flags,
+				      GError **error)
+{
+	FuAlgoltekUsbDevice *self = FU_ALGOLTEK_USB_DEVICE(device);
+	g_autoptr(GBytes) blob_ISP = NULL;
+	g_autoptr(GBytes) blob_payload = NULL;
+	g_autoptr(GByteArray) fw_ISP = g_byte_array_new();
+	g_autoptr(GByteArray) fw_payload = g_byte_array_new();
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 15, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 95, NULL);
+
+	if (!algoltek_device_en(self, error)) {
+		g_prefix_error(error, "ERR: ALGOLTEK_EN failure");
+		return FALSE;
+	}
+
+	if (!algoltek_device_rst(self, 2, error)) {
+		g_prefix_error(error, "ERR: ALGOLTEK_RST failure");
+		return FALSE;
+	}
+
+	fu_device_sleep(FU_DEVICE(self), 900);
+
+	if (!algoltek_device_wrr(self, 0x80AD, 0, error)) {
+		g_prefix_error(error, "ERR: ALGOLTEK_WRR failure");
+		return FALSE;
+	}
+
+	algoltek_device_wrr(self, 0x80C0, 0, error);
```

Comment:
if you're sending error, you need to check for failure, e.g. 

    if (!algoltek_device_wrr(self, 0x80C0, 0, error))
        return FALSE;

Reply:
Updated!